### PR TITLE
Adding test suite for testing IBMQJob in isolation.

### PR DESCRIPTION
**Pull request by delapuente**
The test suite does not require a network connection. It fakes the QJOB object and the API behaviour.

This commit partially refactors the `IBMQJob#status()` method to avoid raising exceptions while reading a property.

## Motivation and Context
Current tests behave as integration tests so, in case of failure, it is not guaranteed the problem arises from IBMJob-related abstractions malfunctioning.

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: test/python/test_ibmqjob_states.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+        {'status': 'COMPLETED'}
+    ]
+
+
+class UnableToInitializeAPI(BaseFakeAPI):
+    """Class for emulating an API unable of initializing."""
+
+    def run_job(self, *_args, **_kwargs):
+        time.sleep(0.2)
+        return {'error': 'ivalid test qobj'}
```
Comment (ajavadia) [inline test/python/test_ibmqjob_states.py]: typo here "invalid"

File: qiskit/backends/ibmq/ibmqjob.py [hunk 2]
```diff
@@ -60,23 +60,19 @@ def __init__(self, q_job, api, is_device):
...
 
-    def result(self, timeout=None, wait=5):
-        # pylint: disable=arguments-differ
-        while self._status == JobStatus.INITIALIZING:
-            time.sleep(0.1)
-        this_result = self._wait_for_job(timeout=timeout, wait=wait)
+    # pylint: disable=arguments-differ
+    def result(self, timeout=60, wait=5):
+        self._wait_for_initialization(timeout=timeout)
+        result = self._wait_for_job(timeout=timeout, wait=wait)
```
Comment (ewinston) [inline qiskit/backends/ibmq/ibmqjob.py]: I think the expectation for `timeout` will be the maximum time for the `result` call to return. Perhaps we can instead start a clock here (`timeout_time = time.time() + timeout`) and pass `timeout_time` to `_wait_for_initialization` and `_wait_for_job`.

File: qiskit/backends/ibmq/ibmqjob.py [hunk 3]
```diff
@@ -105,50 +102,79 @@ def cancel(self):
...
             self._status = JobStatus.DONE
+
         elif job_result['status'] == 'CANCELLED':
             self._status = JobStatus.CANCELLED
-        elif self.exception:
-            self._status = JobStatus.ERROR
-            if self._future_submit.exception():
-                self._exception = self._future_submit.exception()
-            self._status_msg = str(self.exception)
+
```
Comment (ewinston) [inline qiskit/backends/ibmq/ibmqjob.py]: How does this PR handle exceptions raised by `self._future_submit` which is done async and may occur later?

Comment (delapuente) [reply qiskit/backends/ibmq/ibmqjob.py]: There is an error in my assumptions regarding `_submit()` progressing the job to a final state. [This final assignment](https://github.com/QISKit/qiskit-sdk-py/pull/515/files/a105d5b92c20404a8f8e27bc3bbd027a5aa696d7#diff-1601b1931db3e408a0c7973d5605ff4aR313) happening at the end of the method invalidates de second paragraph of the explanation. I'll fix in the next commit. Thanks for the heads-up.

**Comment by delapuente** [comment]
Everything is fixed now. What about a second review?

File: test/python/test_ibmqjob_states.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+"""IBMQJob states test-suite."""
+
+import unittest
```
Comment (diego-plan9) [inline test/python/test_ibmqjob_states.py]: For the new files, can you adjust the license header so it follows the format recently introduced by #550 ?

```
# -*- coding: utf-8 -*-

# Copyright 2018, IBM.
#
# This source code is licensed under the Apache License, Version 2.0 found in
# the LICENSE.txt file in the root directory of this source tree.

# pylint: disable=something-only-if-there-is-no-other-way

"""
Description of the module ...
"""
from qiskit import ...
```

**Comment by ewinston** [comment]
Since #501 is tagged as "high priority" and it's an earlier PR perhaps it should be the other way around.

**Comment by ajavadia** [comment]
@delapuente is this PR rebased after #501 got merged?

**Comment by ajavadia** [comment]
Does this solve #234?

**Comment by delapuente** [comment]
> @delapuente is this PR rebased after #501 got merged?

Yes

> Does this solve #234?

Not yet. It contributes to that but it is not replacing the original IBMQJob suite and more work is needed in other parts of the library.

File: qiskit/backends/ibmq/ibmqjob.py [hunk 5]
```diff
@@ -127,13 +127,22 @@ def result(self, timeout=None, wait=5):
                 raise IBMQJobError('error submitting job: {}'.format(
                     repr(self._future_submit.exception())))
             time.sleep(0.1)
-        this_result = self._wait_for_job(timeout=timeout, wait=wait)
+        try:
+            this_result = self._wait_for_job(timeout=timeout, wait=wait)
+        except TimeoutError as err:
+            # A timeout error retrieving the results does not imply the job
+            # is failing. The job can be still running.
```
[the remark this answers is not in the document]
Comment (atilag) [reply qiskit/backends/ibmq/ibmqjob.py]: Let's mark this for future improvements... we need to merge this asap.

[the remark this answers is not in the document]
Comment (delapuente) [reply qiskit/backends/ibmq/ibmqjob.py]: See #588 for the follow-up.

[the remark this answers is not in the document]
Comment (ewinston) [reply qiskit/backends/ibmq/ibmqjob.py]: When job cancellation works on all IBM Q systems timeout could use that to cancel the job, which could be considered an error.

**Pull request merged**